graylog2-web-interface/src/components/search/MessageDetail.tsx

```diff
@@ -0,0 +1,193 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+import React from 'react';
+import Immutable from 'immutable';
+
+import { Link } from 'components/common/router';
+import { MessageDetailsDefinitionList, ClipboardButton, Icon, Timestamp } from 'components/common';
+import { ButtonGroup, Col, Label, Row } from 'components/bootstrap';
+import StreamLink from 'components/streams/StreamLink';
+import MessageFields from 'components/search/MessageFields';
+import MessageDetailsTitle from 'components/search/MessageDetailsTitle';
+import Routes from 'routing/Routes';
+import AppConfig from 'util/AppConfig';
+import MessagePermalinkButton from 'views/components/common/MessagePermalinkButton';
+import type { Message } from 'views/components/messagelist/Types';
+import type { Stream } from 'views/stores/StreamsStore';
+import type { Input } from 'components/messageloaders/Types';
+
+const MessageActions = ({ messageIndex, messageId }: { messageIndex: string | undefined, messageId: string }) => (
+  <ButtonGroup className="pull-right">
+    <MessagePermalinkButton messageIndex={messageIndex} messageId={messageId} />
+
+    <ClipboardButton title="Copy ID" bsSize="small" text={messageId} />
+  </ButtonGroup>
+);
+
+const InputName = ({ inputs, inputId }: { inputs: Immutable.Map<string, Input> | undefined, inputId: string }) => {
+  const input = inputs?.get(inputId);
+
+  return input ? <span style={{ wordBreak: 'break-word' }}>{input.title}</span> : <>deleted input</>;
+};
+
+const NodeName = ({ nodes, nodeId }) => {
+  const node = nodes.get(nodeId);
+  let nodeInformation;
+
+  if (node) {
+    const nodeURL = Routes.node(nodeId);
+
+    const nodeContent = (
+      <>
+        <Icon name="fork_right" />
+        &nbsp;
+        <span style={{ wordBreak: 'break-word' }}>{node.short_node_id}</span>&nbsp;/&nbsp;
+        <span style={{ wordBreak: 'break-word' }}>{node.hostname}</span>
+      </>
+    );
+
+    nodeInformation = AppConfig.isCloud()
+      ? nodeContent
+      : <a href={nodeURL}>{nodeContent}</a>;
+  } else {
+    nodeInformation = <span style={{ wordBreak: 'break-word' }}>stopped node</span>;
+  }
+
+  return nodeInformation;
+};
+
+const StreamLinks = ({ messageStreams, streamIds, streams }: {
+  messageStreams: Array<Stream>,
+  streamIds: Immutable.Set<string>,
+  streams: Immutable.Map<string, Stream>
+}) => {
+  if (messageStreams) {
+    return (
+      <>
+        {messageStreams.map((stream) => (<li key={stream.id}><StreamLink stream={stream} /></li>))}
+      </>
+    );
+  }
+
+  return (
+    <>
+      {streamIds.map((id) => {
```

Comment:
Maybe we can improve this a bit:
```suggestion
      {streamIds.map((id) => streams.get(id))
          .filter(stream => !!stream)
          .map(stream => <li key={stream.id}><StreamLink stream={stream} /></li>);
```